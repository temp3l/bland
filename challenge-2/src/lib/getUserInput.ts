import * as readlineSync from 'readline-sync';

/**
 * Collects user input for the number of elements and their values.
 * @returns {number[]} - The array of user-input elements.
 */
function getUserInput(): readonly number[] {
  const numberOfElements = parseInt(
    readlineSync.question('Enter the number of elements (max 10): '),
    10,
  );

  if (
    isNaN(numberOfElements) ||
    numberOfElements < 1 ||
    numberOfElements > 10
  ) {
    console.error('Invalid input for the number of elements.');
    process.exit(1);
  }

  const elements = Array.from({ length: numberOfElements }).reduce(
    (acc: readonly number[], _, index) => {
      const element = parseFloat(
        readlineSync.question(`Enter element #${index + 1}: `),
      );

      if (isNaN(element)) {
        console.error('Invalid input for element. Please enter a number.');
        process.exit(1);
      }

      return [...acc, element];
    },
    [],
  ) as readonly number[];

  return elements;
}

export { getUserInput };
