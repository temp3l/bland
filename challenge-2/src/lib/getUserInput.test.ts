/* eslint-disable functional/no-throw-statements */
/* eslint-disable functional/no-return-void */
import { mocked } from 'jest-mock';
import * as readlineSync from 'readline-sync';

import { getUserInput } from './getUserInput';

jest.mock('readline-sync');

const mockedReadlineSync = mocked(readlineSync);
const mockProcessExit = jest
  .spyOn(process, 'exit')
  .mockImplementation((code) => {
    // Forces the code to throw instead of exit
    throw new Error(`Process.exit(${code})`);
  });

describe('getUserInput', () => {
  beforeEach(() => {
    mockProcessExit.mockClear();
  });

  it('should return an array of user-input elements', () => {
    // first value is the number of elements
    mockedReadlineSync.question.mockReturnValueOnce('4');
    mockedReadlineSync.question.mockReturnValueOnce('5');
    mockedReadlineSync.question.mockReturnValueOnce('2');
    mockedReadlineSync.question.mockReturnValueOnce('3');
    mockedReadlineSync.question.mockReturnValueOnce('3');

    const result = getUserInput();

    expect(result).toEqual([5, 2, 3, 3]);
  });

  it('should exit the program - Invalid input for the number of elements.', () => {
    mockedReadlineSync.question.mockReturnValueOnce('NaN');

    expect(() => getUserInput()).toThrowErrorMatchingSnapshot();
  });

  it('should exit the program - Invalid input for the number of elements.', () => {
    mockedReadlineSync.question.mockReturnValueOnce('2');
    mockedReadlineSync.question.mockReturnValueOnce('100');

    expect(() => getUserInput()).toThrowErrorMatchingSnapshot();
  });
});
