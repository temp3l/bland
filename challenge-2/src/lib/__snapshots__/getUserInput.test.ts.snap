// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getUserInput should exit the program - Invalid input for the number of elements. 1`] = `"Process.exit(1)"`;

exports[`getUserInput should exit the program - Invalid input for the number of elements. 2`] = `"Process.exit(1)"`;
